src/web-features.ts

```diff
@@ -0,0 +1,480 @@
+  }
+}
+
+async function fetchResource(event, did, drl, path, responseCache, options) {
+  const endpoints = await getDwnEndpoints(did);
+  if (!endpoints?.length) {
+    throw new Response(
+      "DWeb Node resolution failed: no valid endpoints found.",
+      { status: 530 }
+    );
+  }
+  for (const endpoint of endpoints) {
+    try {
+      const url = `${endpoint.replace(trailingSlashRegex, "")}/${did}/${path}`;
+      const response = await fetch(url, { headers: event.request.headers });
+      if (response.ok) {
+        const match = await options?.onCacheCheck(event, drl);
+        if (match) {
+          cacheResponse(drl, url, response, responseCache);
+        }
+        return response;
+      }
+      console.log(`DWN endpoint error: ${response.status}`);
+      return new Response("DWeb Node request failed", {
+        status: response.status,
+      });
+    } catch (error) {
+      console.log(`DWN endpoint error: ${error}`);
+      return new Response("DWeb Node request failed: " + error, {
+        status: 500,
+      });
+    }
+  }
+}
+
+async function cacheResponse(drl, url, response, cache) {
+  const clonedResponse = response.clone();
+  const headers = new Headers(clonedResponse.headers);
+  headers.append("dwn-cache-time", Date.now().toString());
+  headers.append("dwn-composed-url", url);
+  const modifiedResponse = new Response(clonedResponse.body, { headers });
+  cache.put(drl, modifiedResponse);
+}
+
+/* Service Worker-based features */
+
+async function installWorker(options: any = {}): Promise<void> {
+  const workerSelf = self as any;
+  try {
+    // Check to see if we are in a Service Worker already, if so, proceed
+    // You can call the activatePolyfills() function in your own worker, or standalone as a root worker
+    if (
+      typeof ServiceWorkerGlobalScope !== "undefined" &&
+      workerSelf instanceof ServiceWorkerGlobalScope
+    ) {
+      workerSelf.skipWaiting();
+      workerSelf.addEventListener("activate", (event) => {
+        // Claim clients to make the service worker take control immediately
+        event.waitUntil(workerSelf.clients.claim());
+      });
+      workerSelf.addEventListener("fetch", (event) => {
+        const match = event.request.url.match(didUrlRegex);
+        if (match) {
+          event.respondWith(handleEvent(event, match[1], match[2], options));
+        }
+      });
+    }
+    // If the code gets here, it is not a SW env, it is likely DOM, but check to be sure
+    else if (globalThis?.navigator?.serviceWorker) {
+      const registration = await navigator.serviceWorker.getRegistration("/");
+      // You can only have one worker per path, so check to see if one is already registered
+      if (!registration) {
+        // @ts-ignore
+        const installUrl =
+          options.path ||
+          (globalThis.document
+            ? document?.currentScript?.src
+            : // TODO: review >>> import.meta?.url);
+              undefined);
```

Comment:
@csuwildcat this is what "fixes" the issue for me and everything works normally after built, so could we get rid of it? or detect which environment this is really needed without breaking it...